modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/TableScanNode.java

```diff
@@ -90,190 +64,33 @@ public TableScanNode(
-    public void closeInternal() {
-        super.closeInternal();
+        boolean roTx = context().transactionTime() != null;
 
-        if (activeSubscription != null) {
-            activeSubscription.cancel();
-
-            activeSubscription = null;
-        }
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    protected void rewindInternal() {
-        requested = 0;
-        waiting = 0;
-        curPartIdx = 0;
-
-        if (activeSubscription != null) {
-            activeSubscription.cancel();
-
-            activeSubscription = null;
-        }
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    public void register(List<Node<RowT>> sources) {
-        throw new UnsupportedOperationException();
-    }
-
-    /** {@inheritDoc} */
-    @Override
-    protected Downstream<RowT> requestDownstream(int idx) {
-        throw new UnsupportedOperationException();
-    }
-
-    private void push() throws Exception {
-        if (isClosed()) {
-            return;
-        }
-
-        checkState();
-
-        if (requested > 0 && !inBuff.isEmpty()) {
-            inLoop = true;
-            try {
-                while (requested > 0 && !inBuff.isEmpty()) {
-                    checkState();
-
-                    RowT row = inBuff.poll();
-
-                    if (filters != null && !filters.test(row)) {
-                        continue;
-                    }
-
-                    if (rowTransformer != null) {
-                        row = rowTransformer.apply(row);
-                    }
-
-                    requested--;
-                    downstream().push(row);
-                }
-            } finally {
-                inLoop = false;
-            }
-        }
-
-        if (requested > 0) {
-            if (waiting == 0 || activeSubscription == null) {
-                requestNextBatch();
-            }
-        }
-
-        if (requested > 0 && waiting == NOT_WAITING) {
-            if (inBuff.isEmpty()) {
-                requested = 0;
-                downstream().end();
-            } else {
-                context().execute(this::push, this::onError);
-            }
-        }
-    }
-
-    private void requestNextBatch() {
-        if (waiting == NOT_WAITING) {
-            return;
-        }
-
-        if (waiting == 0) {
-            // we must not request rows more than inBufSize
-            waiting = inBufSize - inBuff.size();
-        }
-
-        Subscription subscription = this.activeSubscription;
-        if (subscription != null) {
-            subscription.request(waiting);
-        } else if (curPartIdx < parts.length) {
-            if (context().transactionTime() != null) {
-                physTable.scan(parts[curPartIdx++], context().transactionTime(), context().localNode()).subscribe(new SubscriberImpl());
+        for (int p : parts) {
+            Publisher<BinaryRow> pub;
+            if (roTx) {
+                pub = physTable.scan(p, context().transactionTime(), context().localNode());
```

Comment:
I'm not sure it's a good idea to open cursors to all partitions at the same time